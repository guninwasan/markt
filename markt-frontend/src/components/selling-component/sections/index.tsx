export * from "./additiona-details";
export * from "./product-flairs";
export * from "./essential-details";
export * from "./media-details";
export * from "./specs";
