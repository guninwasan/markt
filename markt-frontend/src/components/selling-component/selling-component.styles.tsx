import styled, { keyframes } from "styled-components";

const fadeIn = keyframes`
  from { opacity: 0; transform: translateY(10px); }
  to { opacity: 1; transform: translateY(0); }
`;

const pulse = keyframes`
  0% { box-shadow: 0 0 5px rgba(51, 51, 51, 0.1); }
  50% { box-shadow: 0 0 15px rgba(51, 51, 51, 0.3); }
  100% { box-shadow: 0 0 5px rgba(51, 51, 51, 0.1); }
`;

const SellingFormContainer = styled.div`
  width: 75%;
  min-width: 300px;
  max-width: 1200px;
  margin: 20px auto;
  padding: 30px;
  background-color: #fff;
  border-radius: 12px;
  box-shadow: 0 6px 15px rgba(0, 0, 0, 0.15);
  animation: ${fadeIn} 0.5s ease;
`;

const SectionHeader = styled.h2`
  font-size: 1.8em;
  color: grey;
  padding-bottom: 10px;
  border-bottom: 2px solid #ddd;
  margin-top: 30px;
`;

const FormGroup = styled.div`
  margin-bottom: 25px;
  display: flex;
  flex-direction: column;
`;

const FormRow = styled.div`
  display: flex;
  gap: 20px;
  flex-wrap: wrap;
`;

const Label = styled.label`
  font-weight: bold;
  margin-bottom: 8px;
  color: #333;
`;

const TextInput = styled.input`
  width: 90%;
  padding: 14px;
  font-size: 16px;
  border: 1px solid #ddd;
  border-radius: 8px;
  transition: all 0.3s;
  &:focus {
    outline: none;
    border-color: #0066cc;
    box-shadow: 0 0 5px rgba(0, 102, 204, 0.3);
  }
`;

const TextArea = styled.textarea`
  width: 90%;
  padding: 14px;
  font-size: 16px;
  border: 1px solid #ddd;
  border-radius: 8px;
  min-height: 100px;
  transition: all 0.3s;
  &:focus {
    outline: none;
    border-color: #0066cc;
    box-shadow: 0 0 5px rgba(0, 102, 204, 0.3);
  }
`;

const AddMediaButton = styled.button`
  padding: 12px;
  font-size: 16px;
  font-weight: bold;
  color: #000000;
  border: 2px dashed #000000;
  border-radius: 8px;
  background: #f9f9f9;
  cursor: pointer;
  transition: background-color 0.3s, color 0.3s;
  &:hover {
    background: #eef6ff;
  }
`;

const TagContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  gap: 10px;
  margin-top: 10px;
`;

const Tag = styled.span<{ selected?: boolean }>`
  padding: 8px 16px;
  font-size: 0.9em;
  border-radius: 6px;
  cursor: pointer;
  background-color: ${({ selected }) => (selected ? "#0066cc" : "#f0f0f0")};
  color: ${({ selected }) => (selected ? "#fff" : "#333")};
  transition: background-color 0.3s;
  &:hover {
    background-color: ${({ selected }) => (selected ? "#005bb5" : "#ddd")};
  }
`;

const Button = styled.button`
  display: block;
  width: 100%;
  padding: 16px;
  font-size: 18px;
  font-weight: bold;
  background-color: #333;
  color: #fff;
  border: none;
  border-radius: 8px;
  cursor: pointer;
  animation: ${pulse} 1.5s infinite;
  transition: background-color 0.3s;
  &:hover {
    background-color: #555;
  }
`;

export {
  SellingFormContainer,
  SectionHeader,
  FormGroup,
  FormRow,
  Label,
  TextInput,
  TextArea,
  AddMediaButton,
  TagContainer,
  Tag,
  Button,
};
