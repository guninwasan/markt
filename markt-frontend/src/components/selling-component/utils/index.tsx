export * from "./upload-image";
export * from "./valid-form-details";
