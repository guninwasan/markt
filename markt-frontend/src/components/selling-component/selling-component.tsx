import { useState } from "react";
import { SellingFormContainer, Button } from "./selling-component.styles";
import {
  EssentialDetails,
  MediaSection,
  ProductFlairs,
  Specifications,
  AdditionalDetails,
} from "./sections";
import { uploadImage, validateFormData } from "./utils";

const initialFormData = {
  title: "",
  price: "",
  description: "",
  negotiable: false,
  condition: "",
  flairs: [],
  media: [],
  brand: "",
  model: "",
  yearOfManufacture: "",
  color: "",
  dimensions: "",
  weight: "",
  material: "",
  batteryLife: "",
  storageCapacity: "",
  additionalFeatures: "",
  additionalDetails: "",
};

const SellingComponent = () => {
  const [formData, setFormData] = useState<any>(initialFormData);
  const [displayImage, setDisplayImage] = useState<File | null>(
    formData.media[0] || null
  );
  const [mediaFiles, setMediaFiles] = useState<File[]>([]);
  const [priceError, setPriceError] = useState<string>("");

  const handleChange = (
    e: React.ChangeEvent<
      HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement
    >
  ) => {
    const { name, value, type } = e.target;
    if (type === "checkbox") {
      const { checked } = e.target as HTMLInputElement;
      setFormData((prevData: any) => ({
        ...prevData,
        [name]: checked,
      }));
    } else if (type === "file") {
      const { files } = e.target as HTMLInputElement;
      if (files) {
        setFormData((prevData: any) => ({
          ...prevData,
          media: [...prevData.media, ...Array.from(files)],
        }));
      }
    } else {
      setFormData((prevData: any) => ({
        ...prevData,
        [name]: value,
      }));
    }
  };

  const handleTagSelection = (tag: string) => {
    setFormData((prevData: any) => ({
      ...prevData,
      flairs: prevData.flairs.includes(tag)
        ? prevData.flairs.filter((flair: any) => flair !== tag)
        : [...prevData.flairs, tag],
    }));
  };

  const handlePriceChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;
    const priceAsNumber = parseFloat(value);
    if (!isNaN(priceAsNumber) && priceAsNumber >= 0) {
      setPriceError("");
      setFormData((prevData: any) => ({ ...prevData, price: value }));
    } else {
      setPriceError("Please enter a valid positive number for price.");
    }
  };

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();

    const errors = validateFormData(formData, displayImage);
    if (errors.length > 0) {
      alert(errors.join("\n"));
      return;
    }

    try {
      const displayMediaUrl = await uploadImage(displayImage as File);
      const getMediaURLs = await Promise.all(
        mediaFiles.map(async (file: File) => {
          return await uploadImage(file);
        })
      );
    } catch (error) {
      console.error("Error uploading images:", error);
    }
  };

  return (
    <SellingFormContainer>
      <h1>List Your Product</h1>
      <form onSubmit={handleSubmit}>
        <EssentialDetails
          formData={formData}
          handleChange={handleChange}
          handlePriceChange={handlePriceChange}
          priceError={priceError}
        />
        <MediaSection
          setMediaFiles={setMediaFiles}
          mediaFiles={mediaFiles}
          displayImage={displayImage}
          setDisplayImage={setDisplayImage}
        />

        <ProductFlairs
          formData={formData}
          handleTagSelection={handleTagSelection}
        />
        <Specifications formData={formData} handleChange={handleChange} />
        <AdditionalDetails formData={formData} handleChange={handleChange} />
        <Button type="submit">Post Listing</Button>
      </form>
    </SellingFormContainer>
  );
};

export { SellingComponent };
