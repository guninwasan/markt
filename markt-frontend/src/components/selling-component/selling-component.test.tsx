import React from "react";
import { render } from "@testing-library/react";
import { SellingComponent } from "./selling-component";

jest.mock("./sections", () => ({
  EssentialDetails: () => <div>EssentialDetails Component</div>,
  MediaSection: () => <div>MediaSection Component</div>,
  ProductFlairs: () => <div>ProductFlairs Component</div>,
  Specifications: () => <div>Specifications Component</div>,
  AdditionalDetails: () => <div>AdditionalDetails Component</div>,
}));

describe("SellingComponent", () => {
  it("should render all sections", () => {
    const { getByText } = render(<SellingComponent />);
    expect(getByText("EssentialDetails Component")).toBeInTheDocument();
    expect(getByText("MediaSection Component")).toBeInTheDocument();
    expect(getByText("ProductFlairs Component")).toBeInTheDocument();
    expect(getByText("Specifications Component")).toBeInTheDocument();
    expect(getByText("AdditionalDetails Component")).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    const { container } = render(<SellingComponent />);
    expect(container).toMatchSnapshot();
  });
});
