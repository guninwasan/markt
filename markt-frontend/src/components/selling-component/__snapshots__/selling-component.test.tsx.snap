// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SellingComponent should match snapshot 1`] = `
<div>
  <div
    class="sc-blHHSb fuksRV"
  >
    <h1>
      List Your Product
    </h1>
    <form>
      <div>
        EssentialDetails Component
      </div>
      <div>
        MediaSection Component
      </div>
      <div>
        ProductFlairs Component
      </div>
      <div>
        Specifications Component
      </div>
      <div>
        AdditionalDetails Component
      </div>
      <button
        class="sc-keTIit iHiUiW"
        type="submit"
      >
        Post Listing
      </button>
    </form>
  </div>
</div>
`;
